References to linter rule information

- Bash: Shellcheck: `https://github.com/koalaman/shellcheck/wiki/<code>`
  Example: https://github.com/koalaman/shellcheck/wiki/SC2213
- C: CPPCheck: No documentation available
- Java: Checkstyle: `https://checkstyle.sourceforge.io/apidocs/index.html?<code>.html`
  Example: https://checkstyle.sourceforge.io/apidocs/index.html?com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.html
- JavaScript: ESLint: `https://eslint.org/docs/rules/<code>`
  Example: https://eslint.org/docs/rules/for-direction
- Kotlin: KTLint: No specific page for a rule: https://ktlint.github.io/#rules
- Haskell: HLint: No specific page for a rule: https://github.com/ndmitchell/hlint/blob/master/hints.md
- Python: PYLint: No specific page for a rule: https://docs.pylint.org/en/1.6.0/features.html

--- a/docker/dodona-tested.dockerfile
+++ b/docker/dodona-tested.dockerfile
@@ -11,7 +11,7 @@ RUN chmod a+x "$SDKMAN_DIR/bin/sdkman-init.sh"
 RUN /bin/bash -c "source \"$SDKMAN_DIR/bin/sdkman-init.sh\" && sdk install kotlin 1.4.10"
 ENV PATH $SDKMAN_DIR/candidates/kotlin/current/bin:$PATH
 
-RUN pip install jsonschema psutil mako pydantic==1.7.3 toml typing_inspect pylint esprima==4.0.1 lark==0.10.1 pyyaml==5.3.1 python-i18n==0.3.9
+RUN pip install jsonschema psutil mako pydantic==1.7.3 toml typing_inspect pylint esprima==4.0.1 lark==0.10.1 pyyaml==5.3.1 python-i18n==0.3.9 Pygments==2.7.4
 RUN cabal update && cabal install aeson --global --force-reinstalls
 RUN npm install -g eslint@7.23.0
 

--- a/tested/internationalization/en.yaml
+++ b/tested/internationalization/en.yaml
@@ -95,6 +95,8 @@ en:
     generator:
       missing:
         input: "No input found."
+    linter:
+      more-info: "more information"
     python:
       linter:
         crashed: "Pylint crashed"

--- a/tested/internationalization/nl.yaml
+++ b/tested/internationalization/nl.yaml
@@ -95,6 +95,8 @@ nl:
     generator:
       missing:
         input: "Geen invoer gevonden."
+    linter:
+      more-info: "meer informatie"
     python:
       linter:
         crashed: "Pylint gecrasht"

--- a/tested/languages/bash/linter.py
+++ b/tested/languages/bash/linter.py
@@ -67,9 +67,11 @@ def run_shellcheck(bundle: Bundle, submission: Path, remaining: float,
         if not text and code is None:
             continue
         elif not text:
-            text = f"(code {code})"
+            text = f'(code <a href="https://github.com/koalaman/shellcheck/wiki/' \
+                   f'SC{code}" target="_blank">{code}</a>)'
         elif code is not None:
-            text = f"{text} (code {code})"
+            text = f'{text} (code <a href="https://github.com/koalaman/' \
+                   f'shellcheck/wiki/SC{code}" target="_blank">{code}</a>)'
         annotations.append(AnnotateCode(
             row=max(int(shellcheck_object.get('line', "-1")) - 1, 0),
             text=text,

--- a/tested/languages/haskell/linter.py
+++ b/tested/languages/haskell/linter.py
@@ -67,6 +67,11 @@ def run_hlint(bundle: Bundle, submission: Path, remaining: float) \
         hint = hlint_message.get('hint', None)
         if not hint:
             continue
+
+        more_info = get_i18n_string("languages.linter.more-info")
+        hint = f'{hint} <a href="https://github.com/ndmitchell/hlint/blob/master/' \
+               f'hints.md" target="_blank">({more_info})</a>'
+
         hint_from = hlint_message.get('from', None)
         hint_to = hlint_message.get('to', None)
         if hint_from and hint_to:

--- a/tested/languages/java/linter.py
+++ b/tested/languages/java/linter.py
@@ -72,6 +72,12 @@ def run_checkstyle(bundle: Bundle, submission: Path, remaining: float) \
             message = error_element.attrib.get('message', None)
             if not message:
                 continue
+            source = error_element.attrib.get('source', None)
+            if source:
+                more_info = get_i18n_string("languages.linter.more-info")
+                message += f' (https://checkstyle.sourceforge.io/apidocs/' \
+                           f'index.html?{source.replace(".", "/")}.html" ' \
+                           f'target="_blank">{more_info}</a>)'
             annotations.append(AnnotateCode(
                 row=max(int(error_element.attrib.get('line', "-1")) - 1, 0),
                 text=message,

--- a/tested/languages/javascript/linter.py
+++ b/tested/languages/javascript/linter.py
@@ -62,6 +62,10 @@ def run_eslint(bundle: Bundle, submission: Path, remaining: float) \
             text = message.get('message', None)
             if not text:
                 continue
+            rule_id = message.get('ruleId')
+            if rule_id:
+                text += f' (<a href="https://eslint.org/docs/rules/{rule_id}" ' \
+                        f'target="_blank">{rule_id}</a>)'
             annotations.append(AnnotateCode(
                 row=max(int(message.get('line', "-1")) - 1, 0),
                 text=text,

--- a/tested/languages/kotlin/linter.py
+++ b/tested/languages/kotlin/linter.py
@@ -83,6 +83,12 @@ def run_ktlint(bundle: Bundle, submission: Path, remaining: float) \
             message = error.get('message', None)
             if not message:
                 continue
+            rule = error.get('rule', None)
+            if rule:
+                more_info = get_i18n_string("languages.linter.more-info")
+                message += f'({rule}, <a href="https://ktlint.github.io/#rules" ' \
+                           f'target="_blank">{more_info}</a>)'
+
             annotations.append(AnnotateCode(
                 row=max(int(error.get('line', "-1")) - 1, 0),
                 text=message,

--- a/tested/languages/python/linter.py
+++ b/tested/languages/python/linter.py
@@ -72,14 +72,20 @@ def run_pylint(bundle: Bundle, submission: Path, remaining: float) \
         logger.debug("Handling message %s", str(message))
         message_id = message.get('message-id', None)
         message_text = message.get('message', None)
+        more_info = get_i18n_string("languages.linter.more-info")
         if not message_id and not message_text:
             continue
         elif not message_id:
             text = message_text
         elif not message_text:
-            text = f"({message_id})"
+            text = f'({message_id}, <a href="https://pylint.pycqa.org/en/latest/' \
+                   f'technical_reference/features.html#basic-checker-messages" ' \
+                   f'target="_blank">{more_info}</a>)'
         else:
-            text = f"{message_text} ({message_id})"
+            text = f'{message_text} ({message_id},' \
+                   f'<a href="https://pylint.pycqa.org/en/latest/' \
+                   f'technical_reference/features.html#basic-checker-messages" ' \
+                   f'target="_blank">{more_info}</a>)'
         annotations.append(AnnotateCode(
             row=max(int(message.get("line", "-1")) - 1, 0),
             column=max(int(message.get("column", "-1")) - 1, 0),

